pagerduty/resource_pagerduty_extension.go

```diff
@@ -0,0 +1,139 @@
+package pagerduty
+
+import (
+	"log"
+
+	"fmt"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/heimweh/go-pagerduty/pagerduty"
+)
+
+func resourcePagerDutyExtension() *schema.Resource {
+	return &schema.Resource{
+		Create: resourcePagerDutyExtensionCreate,
+		Read:   resourcePagerDutyExtensionRead,
+		Update: resourcePagerDutyExtensionUpdate,
+		Delete: resourcePagerDutyExtensionDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourcePagerDutyExtensionImport,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"endpoint_url": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"extension_objects": {
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"extension_schema": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+		},
+	}
+}
+
+func buildExtensionStruct(d *schema.ResourceData) *pagerduty.Extension {
+	Extension := &pagerduty.Extension{
+		Name:        d.Get("name").(string),
+		Type:        "extension",
+		EndpointURL: d.Get("endpoint_url").(string),
+		ExtensionSchema: &pagerduty.ExtensionSchemaReference{
+			Type: "extension_schema_reference",
+			ID:   d.Get("extension_schema").(string),
+		},
+		ExtensionObjects: expandServices(d.Get("extension_objects")),
+	}
+
+	return Extension
+}
+
+func resourcePagerDutyExtensionCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	Extension := buildExtensionStruct(d)
+
+	log.Printf("[INFO] Creating PagerDuty extension %s", Extension.Name)
+
+	Extension, _, err := client.Extensions.Create(Extension)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(Extension.ID)
+
+	return nil
+}
+
+func resourcePagerDutyExtensionRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	log.Printf("[INFO] Reading PagerDuty extension %s", d.Id())
+
+	extension, _, err := client.Extensions.Get(d.Id())
+	if err != nil {
+		return handleNotFoundError(err, d)
+	}
+
+	d.Set("name", extension.Name)
+
+	return nil
+}
+
+func resourcePagerDutyExtensionUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	extension := buildExtensionStruct(d)
+
+	log.Printf("[INFO] Updating PagerDuty extension %s", d.Id())
+
+	if _, _, err := client.Extensions.Update(d.Id(), extension); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourcePagerDutyExtensionDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*pagerduty.Client)
+
+	log.Printf("[INFO] Deleting PagerDuty extension %s", d.Id())
+
+	if _, err := client.Extensions.Delete(d.Id()); err != nil {
+		return err
```

Comment:
Please check the error code here. If it's equivalent to a `404 Not Found` error, it's better to log the error and indicate that because it wasn't found, Terraform is removing the resource from state. 

Ex:

```go
if _, err := client.Extensions.Delete(d.Id()); err != nil {
  if err == client.404Error { // pseudo code 
    log.Printf("[WARN] Extension (%s) not found, removing from state" , d.Id())
    return nil // returning nil from DELETE will remove item from state
  }
  return err
}
```

This allows Terraform to gracefully handle the removal of resources that have been destroyed outside of Terraform.